### PR TITLE
Azure Cloud Service and CB explore fixes

### PR DESCRIPTION
**Pull request by eisber**

### CODE REVIEW

**Comment by JohnLangford** [comment]
1) Need to make tests work.
2) @lhoang29 can you review the C# side?

**Comment by lhoang29** [comment]
looking..

File: cs/azure/Trainer/CountingCheckpointPolicy.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace VowpalWabbit.Azure.Trainer
```
Comment (lhoang29) [inline cs/azure/Trainer/CountingCheckpointPolicy.cs]: looks like this class is duplicated? there's another one in Trainer/Checkpoint/

File: cs/azure/Trainer/LearnEventProcessorFactory.cs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            this.telemetry.Context.Component.Version = GetType().Assembly.GetName().Version.ToString();
+
+            this.evalOperation = new EvalOperation(settings);
+            //this.latencyOperation = new LatencyOperation();
+
+            this.deserializeBlock = new TransformManyBlock<PipelineData, PipelineData>(
+                (Func<PipelineData, IEnumerable<PipelineData>>)this.Deserialize,
+                new ExecutionDataflowBlockOptions
+                {
+                    MaxDegreeOfParallelism = 4, // Math.Max(2, Environment.ProcessorCount - 1),
```
Comment (lhoang29) [inline cs/azure/Trainer/LearnEventProcessorFactory.cs]: should this be turned back on to ProcesscorCount or did you find it troublesome? in the past I found that proc count or proc count * 2 are typically good defaults

File: cs/azure/Trainer/LearnEventProcessorFactory.cs [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+            {
+                this.evalOperation.Dispose();
+                this.evalOperation = null;
+            }
+
+            //if (this.latencyOperation != null)
+            //{
+            //    this.latencyOperation.Dispose();
+            //    this.latencyOperation = null;
+            //}
```
Comment (lhoang29) [inline cs/azure/Trainer/LearnEventProcessorFactory.cs]: remove unused latencyOperation

File: cs/azure/Trainer/LearnEventProcessorFactory.cs (hunk 2 above)
Comment (lhoang29) [inline cs/azure/Trainer/LearnEventProcessorFactory.cs]: Also, just checking, I think the Transform*Block maintains event order yes?

File: cs/azure/Trainer/CountingCheckpointPolicy.cs (hunk 1 above)
Comment (eisber) [inline cs/azure/Trainer/CountingCheckpointPolicy.cs]: excellent point... missing more code now

---
In reply to: [66970805](https://github.com/JohnLangford/vowpal_wabbit/pull/1013#discussion_r66970805) [](ancestors = 66970805)

File: cs/azure/Trainer/LearnEventProcessorFactory.cs (hunk 2 above)
Comment (eisber) [inline cs/azure/Trainer/LearnEventProcessorFactory.cs]: yes

---
In reply to: [66973713](https://github.com/JohnLangford/vowpal_wabbit/pull/1013#discussion_r66973713) [](ancestors = 66973713)

Comment (eisber) [inline cs/azure/Trainer/LearnEventProcessorFactory.cs]: it's less true for the trainer. Let's turn this up once I get the JSON.NET float parser in...

---
In reply to: [66972619](https://github.com/JohnLangford/vowpal_wabbit/pull/1013#discussion_r66972619) [](ancestors = 66972619)

File: cs/azure/Trainer/LearnEventProcessorFactory.cs (hunk 3 above)
Comment (eisber) [inline cs/azure/Trainer/LearnEventProcessorFactory.cs]: I actually want to implement this in the future... just not enough time now

---
In reply to: [66973362](https://github.com/JohnLangford/vowpal_wabbit/pull/1013#discussion_r66973362) [](ancestors = 66973362)

**Comment by JohnLangford** [comment]
What is the status on this?  Can you update to the current master?  I'd like to drop a new version soon.

**Comment by eisber** [comment]
Updated. I think Louie wasn’t quite finished with the Review, but that might take a bit longer.

Markus

**Comment by JohnLangford** [comment]
@lhoang29 What's the status of your look-through?

File: cs/azure/Trainer/OnlineTrainerSettings.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+﻿namespace VowpalWabbit.Azure.Trainer
+{
+    public class OnlineTrainerSettings
+    {
+        // Model blobs
+        public const string ModelContainerName = "mwt-models";
+        public const string LatestModelBlobName = "current";
```
Comment (lhoang29) [inline cs/azure/Trainer/OnlineTrainerSettings.cs]: use the contract library here instead?

File: cs/azure/Trainer/Util/Util.cs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+using System.Buffers;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace VowpalWabbit.Azure.Trainer
+{
+    public class MyArrayPool : IArrayPool<char>
```
Comment (lhoang29) [inline cs/azure/Trainer/Util/Util.cs]: is this class being used? the name looks suspiciously dead...

File: cs/azure/Trainer/Util/Util.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        public void Return(char[] array)
+        {
+            this.pool.Return(array);
+        }
+    }
+
+   
+    public static class Util
+    {
+        public static unsafe bool ReadDoubleString(char[] chars, int offset, int length, out double acc)
```
Comment (lhoang29) [inline cs/azure/Trainer/Util/Util.cs]: is this being tested? also I'm curious what kind of perf gain is obtained here vs the built-in Convert.ToDouble or parse ..

File: cs/azure_service/OnlineTrainerContent/InstallAgent.ps1 [hunk 7]
```diff
@@ -0,0 +1,301 @@
+﻿#########################################################
```
Comment (lhoang29) [inline cs/azure_service/OnlineTrainerContent/InstallAgent.ps1]: is this file auto-generated?

File: cs/azure/Trainer/LearnEventProcessorFactory.cs [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+            this.performanceCounters.Stage1_JSON_Queue.RawValue = this.deserializeBlock.InputCount;
+            this.performanceCounters.Stage2_Learn_Queue.RawValue = this.learnBlock.InputCount;
+            this.performanceCounters.Stage3_Checkpoint_Queue.RawValue = this.checkpointBlock.InputCount;
+        }
+
+        internal async Task Stage0_Split(PartitionContext context, IEnumerable<EventData> messages)
+        {
+            foreach (EventData eventData in messages)
+            {
+                try
```
Comment (lhoang29) [inline cs/azure/Trainer/LearnEventProcessorFactory.cs]: remove noop try-catch

**Comment by lhoang29** [comment]
I finished another pass through, ignoring most VS plumbing files (.config, .proj, etc...).

File: cs/azure/Trainer/OnlineTrainerSettings.cs (hunk 4 above)
Comment (eisber) [inline cs/azure/Trainer/OnlineTrainerSettings.cs]: I want to avoid circular dependencies…

File: cs/azure/Trainer/Util/Util.cs (hunk 5 above)
Comment (eisber) [inline cs/azure/Trainer/Util/Util.cs]: Will be used by the upcoming JSON.NET changes.

File: cs/azure/Trainer/Util/Util.cs (hunk 6 above)
Comment (eisber) [inline cs/azure/Trainer/Util/Util.cs]: I’m still evaluating. If the process is heavily multi-threaded (>16 threads), double.TryParse by requiring a string as input puts heavy pressure on GC. By accessing the char[] direct this is gone.

File: cs/azure_service/OnlineTrainerContent/InstallAgent.ps1 (hunk 7 above)
Comment (eisber) [inline cs/azure_service/OnlineTrainerContent/InstallAgent.ps1]: no

**Pull request merged**